tests/lib/fortigate_test.py

```diff
@@ -0,0 +1,352 @@
+      args = list(args)
+      for field in kwargs['remove_fields']:
+        remove_regex = '.*' + field + '.*'
+        args[1] = re.sub(remove_regex, '', args[1])
+
+      return string.Formatter.format(*args, **kwargs)
+    return string.Formatter.format(*args, **kwargs)
+
+  def get_value(self, key, args, kwds):
+    try:
+      return kwds[key]
+    except KeyError:
+      return self.DEFAULT_VALUES[key]
+
+
+class FortigateTest(unittest.TestCase):
+  def setUp(self):
+    self.naming = mock.create_autospec(naming.Naming)
+
+    def get_addr_side_eff(host):
+      hosts = {
+          'SOME_HOST': [nacaddr.IP('10.0.0.0/8')],
+          'SOME_HOST2': [nacaddr.IP('20.0.0.0/8')]
+      }
+      return hosts[host]
+
+    def get_port_side_eff(*args):
+      hosts = {
+          'HTTP': ['80'],
+          'HTTPS': ['443'],
+          'SSH': ['22'],
+          'WHOIS': ['43']
+      }
+      return hosts[args[0]]
+
+    self.naming.GetNetAddr.side_effect = get_addr_side_eff
+    self.naming.GetServiceByProto.side_effect = get_port_side_eff
+    self.fmt = CustomFormatter()
+
+  def testGoodHeader(self):
+    term = self.fmt.format(TERM_TEMPLATE)
+    acl = fortigate.Fortigate(policy.ParsePolicy(GOOD_HEADER + term,
+                                                 self.naming), EXP_INFO)
+
+    expected_sig = 'edit 2'
+
+    get_net_calls = [mock.call('SOME_HOST')] * 2
+    get_server_by_proto_calls = [mock.call('HTTP', 'tcp')] * 2
+
+    self.assertIn(expected_sig, str(acl), '[%s]' % str(acl))
+    self.naming.GetNetAddr.assert_has_calls(get_net_calls)
+    self.naming.GetServiceByProto.assert_has_calls(get_server_by_proto_calls)
+
+  def testBadHeader(self):
+    term = self.fmt.format(TERM_TEMPLATE)
+    parsed_p = policy.ParsePolicy(BAD_HEADER + term,
+                                  self.naming)
+
+    self.assertRaises(fortigate.FilterError,
+                      fortigate.Fortigate,
+                      parsed_p,
+                      EXP_INFO)
+
+  def testAction(self):
+    accept_term = self.fmt.format(TERM_TEMPLATE, action='accept')
+    deny_term = self.fmt.format(TERM_TEMPLATE, action='deny')
+    reject_term = self.fmt.format(TERM_TEMPLATE, action='reject')
+
+    accept_acl = fortigate.Fortigate(
+        policy.ParsePolicy(GOOD_HEADER + accept_term,
+                           self.naming), EXP_INFO)
+    deny_acl = fortigate.Fortigate(
+        policy.ParsePolicy(GOOD_HEADER + deny_term,
+                           self.naming), EXP_INFO)
+    reject_acl = fortigate.Fortigate(
+        policy.ParsePolicy(GOOD_HEADER + reject_term,
+                           self.naming), EXP_INFO)
+
+    accept_sig = 'set action accept'
+    deny_sig = 'set action deny'
+    self.assertIn(
+        accept_sig, str(accept_acl), '[%s]' % str(accept_acl))
+    self.assertIn(
+        deny_sig, str(deny_sig), '[%s]' % str(deny_acl))
+    self.assertIn(
+        deny_sig, str(reject_acl), '[%s]' % str(reject_acl))
+
+  def testAddresses(self):
```

Comment:
You are collapsing a lot of checks into a single test with multiple unrelated asserts.
The output on a failure looks like the following

```
Traceback (most recent call last):
  File "tests/lib/fortigate_test.py", line 156, in testAction
    self.assertIn(
AssertionError: 'set action accept' not found in 'config firewall address\n\tedit 10.0.0.0/8\n\t\tset subnet 10.0.0.0/8\n\tnext\n \nend\n \nconfig firewall service custom\n \nend\n \nconfig firewall policy\n\tedit 2\n\t\tset comments good-term-2\n\t\tset srcintf wan1\n\t\tset dstintf wan2\n\t\tset dstaddr 10.0.0.0/8\n\t\tset srcaddr 10.0.0.0/8\n\t\tset action aaccept\n\t\tset service HTTP\n\t\tset schedule always\n\t\tset logtraffic all\n\tnext\n\nend' : [config firewall address
	edit 10.0.0.0/8
		set subnet 10.0.0.0/8
	next

end


end

config firewall service custom

end


end

config firewall policy
	edit 2
		set comments good-term-2
		set srcintf wan1
		set dstintf wan2
		set dstaddr 10.0.0.0/8
		set srcaddr 10.0.0.0/8
		set action aaccept
		set service HTTP
		set schedule always
		set logtraffic all
	next

end]

----------------------------------------------------------------------
Ran 9 tests in 1.024s
```

Execution inside of a single test also stops at the first failure. If you had multiple issues such as both accept and deny failing for different reasons, this test would not report that both are wrong.

You may wish to look into paramaterization as this feels like what you are wanting to do. There has been talk about redoing a lot of our tests to use that.